src/stream.rs

```diff
@@ -0,0 +1,104 @@
+use crate::device::Device;
+use crate::utils::mlx_describe;
+
+/// Parameter type for all MLX operations.
+///
+/// Use this to control where operations are evaluated:
+///
+/// If omitted it will use the [default()], which will be [Device::gpu()] unless
+/// set otherwise.
+pub struct StreamOrDevice {
+    stream: Stream,
+}
+
+impl StreamOrDevice {
+    #[allow(dead_code)]
+    fn new(stream: Stream) -> StreamOrDevice {
+        StreamOrDevice { stream }
+    }
+
+    /// The default stream on the default device.
+    ///
+    /// This will be [Device::gpu()] unless [Device::set_default()]
+    /// sets it otherwise.
+    pub fn default() -> StreamOrDevice {
+        StreamOrDevice {
+            stream: Stream::new(),
+        }
+    }
+
+    pub fn new_with_device(device: &Device) -> StreamOrDevice {
+        StreamOrDevice {
+            stream: Stream::default_stream(device),
+        }
+    }
+
+    /// The `[Stream::default_stream()] on the [Device::cpu()]
+    pub fn cpu() -> StreamOrDevice {
+        StreamOrDevice {
+            stream: Stream::default_stream(&Device::cpu()),
+        }
+    }
+
+    /// The `[Stream::default_stream()] on the [Device::gpu()]
+    pub fn gpu() -> StreamOrDevice {
+        StreamOrDevice {
+            stream: Stream::default_stream(&Device::gpu()),
+        }
+    }
+
+    #[allow(dead_code)]
+    fn get_ctx(&self) -> mlx_sys::mlx_stream {
+        self.stream.c_stream
+    }
+}
+
+impl std::fmt::Display for StreamOrDevice {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", self.stream)
+    }
+}
+
+/// A stream of evaluation attached to a particular device.
+///
+/// Typically, this is used via the `stream:` parameter on a method with a [StreamOrDevice]:
+pub struct Stream {
+    c_stream: mlx_sys::mlx_stream,
+}
+
+impl Stream {
+    fn new_with_mlx_mlx_stream(stream: mlx_sys::mlx_stream) -> Stream {
+        Stream { c_stream: stream }
+    }
+
+    pub fn new() -> Stream {
```

Comment:
I think `clippy` would suggest impl the `Default` trait while keeping the `new()` method